feat(react-graphql-mock): mock graphql queries and mutations with miragejs

wan54 opened the pull request:
## Checklist
- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [x] Docs related to the changes have been added/updated (if applicable)

wan54 commented:
https://miragejs.com/

Provides:

- benefits that you already know offered by a mock server plus
- ability to mock different scenarios by implementing logic that response to different data driven by request parameters/payloads

gabe-lyons commented on datahub-web-react/src/graphql-mock/fixtures/browseDataset/datasetBrowseKafka.ts:
```diff
@@ -0,0 +1,207 @@
+                count: 2,
+                total: 2,
+                metadata: {
+                    path: ['prod', 'kafka', 'wah', 'corollary'],
```
@wan54  can you make this sample data a little more realistic? I worry someone might boot up this mock server and be confused when they see words like this.

gabe-lyons replied:
I know http://demo.datahubproject.io/ had words like this in certain instances but we removed them due to them being confusing.

gabe-lyons commented on datahub-web-react/src/graphql-mock/fixtures/browseDataset/datasetBrowseSnowflake.ts:
```diff
@@ -0,0 +1,101 @@
+            count: 0,
+            total: 0,
+            metadata: {
+                path: ['prod', 'snowflake', 'novelty'],
```
again, we had some randomly generated datasets in the demo data but I found they caused more confusion than they helped. Let's use more realistic names in this sample data.

gabe-lyons commented on datahub-web-react/src/graphql-mock/fixtures/searchResult/datasetSearchResult.ts:
```diff
@@ -0,0 +1,138 @@
+        },
+        {
+            platform: 'bigquery',
+            paths: [
+                { path: 'wah.hitch', count: 2 },
+                { path: 'wah.agnes', count: 1 },
+                { path: 'twine.hitch', count: 1 },
+            ],
```
same comment here re: realism of sample data

gabe-lyons reviewed (commented):
This looks great @wan54 - thank you for putting this contribution together. A few questions:

1) How much of this code will be included in the bundle that gets shipped to users who aren't running in mock mode? I want to make sure this change doesn't enlarge our bundle size too much.

2) can you show some examples of how these mocks are used in tests so folks can start using this mocking system for testing going forward? Migrating some existing tests to this mock system would be very illustrative.

3) Do you have any thoughts on the plan to maintain these mocks? Our graph is extended not too infrequently. Any tips on how to identify gaps in mocks when they appear? What sort of gaps in the mocks would cause serious issues vs which gaps would fail gracefully?

wan54 commented:
Thanks @gabe-lyons. I'll work on making the names more readable soon as I get the chance.

> This looks great @wan54 - thank you for putting this contribution together. A few questions:
> 
> 1. How much of this code will be included in the bundle that gets shipped to users who aren't running in mock mode? I want to make sure this change doesn't enlarge our bundle size too much.

None. There is a guard before creating the mock server so that when the build is executed the mock server and everything related to it won't get imported by webpack. Please see in the following source:

- https://github.com/linkedin/datahub/pull/2548/commits/1ae67055ce80e8269bc9eb715d73669824a16898#diff-2fb968e31739b9f47081ccc9d94ca1d89f98207fef9b8bff65d5b7a3d5650f0dR4
- https://github.com/linkedin/datahub/pull/2548/commits/1ae67055ce80e8269bc9eb715d73669824a16898#diff-da3686cd43964912b03f3f443d0b7f2b4842defdfb0057a990f9fbdb51316832R75

> 2. can you show some examples of how these mocks are used in tests so folks can start using this mocking system for testing going forward? Migrating some existing tests to this mock system would be very illustrative.

- I'll create another PR with Cypress + MirageJS integration to best show case it. Coming soon after I've addressed the feedback.

> 3. Do you have any thoughts on the plan to maintain these mocks? Our graph is extended not too infrequently. Any tips on how to identify gaps in mocks when they appear? What sort of gaps in the mocks would cause serious issues vs which gaps would fail gracefully?

Yes. Thanks to the schema it is impossible to create mock data and have MirageJS to successfully response to the query or mutation without obeying to the shapes and relationships defined in it. If mock data are not created right Apollo client will throw the error in the browser console and simply follow the stack trace to identify the root cause. 

Also, after the schema or mock data is updated MirageJS server has to be restarted by simply reloading the browser.

wan54 commented:
There is also another option that this PR may be used as reference by the team and they can implement the logic how to mock the data for all the queries and mutations in their own way.

gabe-lyons commented:
Got it- thanks! This lgtm then.

shirshanka commented:
@wan54 : could you see if the build failure is unrelated and just needs you to rebase your fork?

wan54 commented:
It is unrelated. Seems to be caused by the python test failure https://github.com/linkedin/datahub/pull/2548/checks?check_run_id=2659659309#step:7:2634

hsheth2 commented:
@wan54 it seems you rebased on top of a slightly outdated version, and so you're missing the python tests fix that was added in https://github.com/linkedin/datahub/commit/48d2b94203275ea07085f6131fe7ffd3d1011d11. Can you merging/rebasing with latest master?

wan54 commented:
Just rebased

wan54 commented:
If you guys are going to merge this PR, merge https://github.com/linkedin/datahub/pull/2597 instead. It's got changes from this PR + e2e test examples with Cypress.

shirshanka commented:
Closing this one as #2597 is the appropriate PR to track for this work.
